api/src/main/java/org/pmiops/workbench/db/dao/DataSetServiceImpl.java

```diff
@@ -664,14 +664,58 @@ public static String capitalizeFirstCharacterOnly(String text) {
     return StringUtils.capitalize(text.toLowerCase());
   }
 
+  private static String generateSqlWithEnvironmentVariables(
+      String query, KernelTypeEnum kernelTypeEnum) {
+    switch (kernelTypeEnum) {
+      case PYTHON:
+        return query.replaceAll(
+            "\\$\\{projectId}.\\$\\{dataSetId}", "\"\"\" + os.environ[\"WORKSPACE_CDR\"] + \"\"\"");
+      case R:
+        return query.replaceAll(
+            "\\$\\{projectId}.\\$\\{dataSetId}", "\", Sys.getenv(\"WORKSPACE_CDR\"), \"");
+      default:
+        return query;
+    }
+  }
+
+  private static String fillInQueryParams(
+      String query, Map<String, QueryParameterValue> queryParameterValueMap) {
+    StringBuilder stringBuilder = new StringBuilder(query);
+    queryParameterValueMap.forEach(
+        (key, value) -> {
+          if (StandardSQLTypeName.ARRAY.equals(value.getType())) {
+            String stringToReplace = "unnest(@".concat(key.concat(")"));
+            int startingIndex = stringBuilder.indexOf(stringToReplace);
+            stringBuilder.replace(
+                startingIndex,
+                startingIndex + stringToReplace.length(),
+                "("
+                    .concat(
+                        nullableListToEmpty(value.getArrayValues()).stream()
+                            .map(QueryParameterValue::getValue)
+                            .collect(Collectors.joining(", ")))
+                    .concat(")"));
+          } else {
+            String stringToReplace = "@".concat(key);
+            int startingIndex = stringBuilder.indexOf(stringToReplace);
+            stringBuilder.replace(
+                startingIndex,
+                startingIndex + stringToReplace.length(),
+                Optional.ofNullable(value.getValue()).orElse(""));
```

Comment:
nit: I think this reads better with `Optional.ofNullable(...).ifPresent(v -> stringBuilder.replace(..., v)`

But also, I'm not used to seeing replace used on a stringBuilder. It's generally expected that you build up the correct string from the beginning. Is there a way to do the swapping earlier?

Reply:
Not without changing the Cohort Builder SQL building code, unfortunately. Will change to the Optional.ofNullable method though.

Reply:
OK, I really want to get to tinkering on a new third party query builder.